src/npcmove.cpp

```diff
@@ -399,20 +404,141 @@ std::vector<sphere> npc::find_dangerous_explosives() const
+    const item &candidate_weap = candidate.get_wielded_item() ? *candidate.get_wielded_item() :
+                                 null_item_reference();
+    double candidate_weap_val = candidate.weapon_value( candidate_weap );
+    float candidate_health =  candidate.hp_percentage() / 100.0f;
+    float armour = estimate_armour( candidate );
+    float speed = std::max( 0.25f, candidate.get_speed() / 100.0f );
+    if( !my_gun ) {
+        speed = std::max( speed, 0.5f );
+    };
+
+    threat += my_gun && enemy ? candidate.get_dodge() / 2.0f : candidate.get_dodge();
+    threat += armour;
+    add_msg_debug( debugmode::DF_NPC_COMBATAI, "%s assesses %s defense value as %1.2f.",
+                   name, candidate.disp_name( true ), threat );
+
+    if( enemy && candidate_gun && !my_gun ) {
+        add_msg_debug( debugmode::DF_NPC_COMBATAI,
+                       "%s has a gun and %s doesn't; %s adjusts threat accordingly.",
+                       candidate.disp_name(), name, name );
+        candidate_weap_val *= 1.5f;
+    }
+    threat += candidate_weap_val;
+    add_msg_debug( debugmode::DF_NPC_COMBATAI, "%s assesses %s weapon value as %1.2f.",
+                   name, candidate.disp_name( true ), candidate_weap_val );
+
+    if( enemy ) {
+        threat -= static_cast<float>( personality.aggression );
+    } else {
+        threat +=  static_cast<float>( personality.bravery );
+    }
+
+    threat *= speed;
+    add_msg_debug( debugmode::DF_NPC_COMBATAI, "%s scales %s threat by %1.0f%% based on speed.",
+                   name, candidate.disp_name( true ), speed * 100.0f );
+    threat *= candidate_health;
+    add_msg_debug( debugmode::DF_NPC_COMBATAI,
+                   "%s scales %s threat by %1.0f%% based on remaining health.", name,
+                   candidate.disp_name( true ), candidate_health * 100.0f );
+
+    float perception_fuzz = threat * rng( -1.0f,
+                                          1.0f ) * std::max( ( 20.0f - static_cast<float>( get_per() ) ), 0.0f ) / 20.0f;
+    threat += perception_fuzz;
+
+    add_msg_debug( debugmode::DF_NPC_COMBATAI, "%s randomizes %s threat by %1.1f%% based on perception",
+                   name, candidate.disp_name( true ), perception_fuzz );
+    add_msg_debug( debugmode::DF_NPC, "%s assesses %s final threat as %1.2f", name,
+                   candidate.disp_name( true ),
+                   threat );
+    return std::min( threat, NPC_CHARACTER_DANGER_MAX );
+}
+
+float npc::evaluate_self( bool my_gun ) const
+{
+    float threat = 0.0f;
+    const double &my_weap_val = ai_cache.my_weapon_value;
+    // the worse pain the NPC is in, the more likely they are to overestimate the severity of their injuries.
+    // the more perceptive they are, the more they're able to see how bad it really is.
+    float pain_factor = rng( 0.0f,
+                             static_cast<float>( get_pain() ) / static_cast<float>( get_per() ) );
```

Comment:
```suggestion
                             static_cast<float>( get_pain() ) / std::max( 1.0, static_cast<float>( get_per() ) ) );
```

You may want to add a `std::max` here to avoid a potential division by zero.

Reply:
oh yes, I noticed that and then forgot, thanks for reminding me.

Reply:
oh wait it looks like I did do that. Thakns anwyyay